File: C#/Binary Search Tree/450. Delete Node in a BST.cs
/* 450. Delete Node in a BST
Difficulty: Medium
Submission Link: https://leetcode.com/problems/delete-node-in-a-bst/submissions/1384565498/ */


public class Solution {
    public TreeNode DeleteNode(TreeNode root, int key) {
        if(root == null)
        {
            return null;
        }

        if(root.val<key)
        {
            root.right = DeleteNode(root.right, key);
        }
        else if(root.val>key)
        {
            root.left = DeleteNode(root.left, key);
        }
        else
        {
            if(root.left == null && root.right==null)
            {
                return null;
            }
            else if(root.left == null)
            {
                return root.right;
            }
            else if(root.right == null)
            {
                return root.left;
            }
            else
            {
                var temp = root.right;
                while(temp.left!=null)
                {
                    temp = temp.left;
                }

                root.val = temp.val;
                root.right = DeleteNode(root.right, temp.val);
            }
        }

        return root;
    }
}

File: Python3/Binary Search Tree/450. Delete Node in a BST.py
""" 450. Delete Node in a BST
Difficulty: Medium
Submission Link: https://leetcode.com/problems/delete-node-in-a-bst/submissions/1384563835/ """

class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if root:
            if key < root.val:
                root.left = self.deleteNode(root.left, key)
            elif key > root.val:
                root.right = self.deleteNode(root.right, key)
            else:
                if not root and not root.right:
                    return None
                if not root.left or not root.right:
                    return root.left if root.left is not None else root.right
                temp = root.left
                while temp.right:
                    temp = temp.right
                root.val = temp.val
                root.left = self.deleteNode(root.left, temp.val)
        return root